test/dsp_test.exs

```diff
@@ -0,0 +1,81 @@
+defmodule Test.Dsp do
+  use ExUnit.Case, async: true
+
+  alias Test.Support.Utils
+  alias ExUid2.Dsp
+
+  test "When started, the dsp client can decrypt a token" do
+    now = :os.system_time(:millisecond)
+    opts = make_opts(now)
+
+    prepare_dsp(opts)
+    Dsp.refresh()
+
+    keyring = Utils.keyring(opts)
+    token = make_token(keyring)
+
+    {:ok, expected_uid2} =
+      token
+      |> :base64.decode()
+      |> ExUid2.Encryption.decrypt_v2_token(keyring, now)
+
+    {:ok, uid2} = Utils.try_until_success(fn -> Dsp.decrypt_token(token) end)
+
+    assert match?(^expected_uid2, uid2)
+  end
+
+  test "When started, the dsp client will return an error if an attempt to decrypt the token is made before the keyring is fetched" do
+    now = :os.system_time(:millisecond)
+    opts = make_opts(now)
+
+    prepare_dsp(opts)
+
+    keyring = Utils.keyring(opts)
+    token = make_token(keyring)
+
+    result = Dsp.decrypt_token(token)
+
+    assert match?({:error, :no_keyring_stored}, result)
+  end
+
+  test "Failed attempts to call the UID2 operator server will be retried" do
+    now = :os.system_time(:millisecond)
+    opts = make_opts(now)
+
+    prepare_dsp(opts)
+
+    Req.Test.expect(ExUid2.Dsp, &Req.Test.transport_error(&1, :econnrefused))
+    Req.Test.expect(ExUid2.Dsp, &Plug.Conn.send_resp(&1, 500, "internal server error"))
```

Comment:
Do I understand correctly that this means the third request will succeed?

Reply:
That's correct.